### PR TITLE
Speed up nftables via batching

### PR DESCRIPTION
**Pull request by sbs2001**
Fixes https://github.com/crowdsecurity/cs-firewall-bouncer/issues/114 .

The `backend` interface has one more method `Commit` . This method is supposed to be called to actually commit changes to the system, instead of relying `Add` and `Delete` methods. Only nftables currently follows this while other modes still work. 

`nftables` now instead of calling `Flush` for each `Add` or `Delete` does the following:

1. Compare decision stream with current state. Decisions are filtered off if their ip is already present (for new decision) or if ip is absent(decision is expired)
2. The filtered decisions are then batched into size of `200` (arbitary for now TBD) and `Flush` is called only once per batch

Benchmark : 
Able to insert 100k decisions in ~45s

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -158,6 +158,11 @@ func main() {
 					noun = "decision"
 				}
 				if nbDeletedDecisions > 0 {
+					log.Info("commiting expired decisions")
```
Comment (buixor) [inline main.go:161]: debug maybe ?

File: main.go [hunk 2]
```diff
@@ -158,6 +158,11 @@ func main() {
...
+					if err := backend.Commit(); err != nil {
+						log.Errorf("unable to commit delete decisions %v", err)
+					}
+					log.Info("commited expired decisions")
```
Comment (buixor) [inline main.go:165]: debug maybe ?

File: main.go [hunk 3]
```diff
@@ -180,6 +185,11 @@ func main() {
 					noun = "decision"
 				}
 				if nbNewDecisions > 0 {
+					log.Info("commiting added decisions")
```
Comment (buixor) [inline main.go:188]: same x2

**Review by buixor — CHANGES_REQUESTED** [review]
minor logging

**Comment by jarppiko** [comment]
Hi @sbs2001 is it necessary to remove the duplicates or is it an additional feature?

**Comment by sbs2001** [comment]
@jarppiko  nftables doesn't permit inserting duplicate IPs in single batch. It also errors if the item being inserted is already present in set.

**Comment by jarppiko** [comment]
> @jarppiko nftables doesn't permit inserting duplicate IPs in single batch. It also errors if the item being inserted is already present in set.

This issue with importing duplicates has been partly fixed in `nftables`. Importing duplicates in a single batch does not cause errors with recent `nftables`, but if there is an existing rule, it does not update it. 

```
# cat ip-blocklist-duplicate-test.csv
value
1.116.1.107
1.116.1.22
1.116.1.239
1.116.101.225
1.116.102.149
1.116.1.107
1.116.101.225
1.116.101.225
# cscli decisions import  --reason "Test" --duration 3h -i ip-blocklist-duplicate-test.csv
INFO[01-02-2022 02:19:03 PM] 8 decisions successfully imported
#
```

And `var/log/crowdsec-firewall-bouncer.log` shows:  
```
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.1.107' for '2h59m56.957743109s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.1.22' for '2h59m56.957741101s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.1.239' for '2h59m56.957740476s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.101.225' for '2h59m56.957740058s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.102.149' for '2h59m56.957739793s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.1.107' for '2h59m56.957739377s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.101.225' for '2h59m56.957739125s'"
time="01-02-2022 14:19:06" level=debug msg="adding '1.116.101.225' for '2h59m56.957738878s'"
time="01-02-2022 14:19:06" level=info msg="8 decisions added"
time="01-02-2022 14:19:06" level=debug msg="committed 8 decisions"
time="01-02-2022 14:19:06" level=info msg="committed 8 added decisions, 0 deleted decisions"
```

So the question is whether this import behavior is for upstream of crowdsec to fix.  What do you guys think?

File: nftables.go [hunk 4]
```diff
@@ -298,71 +295,169 @@ func (n *nft) Init() error {
...
+		if _, ok := vals[*d.Value]; ok {
+			continue
+		}
+		vals[*d.Value] = struct{}{}
+		if _, err := time.ParseDuration(*d.Duration); err != nil {
+			d.Duration = &defaultTimeout
+		}
+		*d.Value = strings.Split(*d.Value, "/")[0]
+		finalDecisions = append(finalDecisions, d)
+	}
```
Comment (jarppiko) [inline nftables.go:455]: Hi @sbs2001, now the logic is that the first decision will be saved, should it actually be the latest decision that gets a place in the normalized array?

File: nftables.go [hunk 5]
```diff
@@ -298,71 +295,169 @@ func (n *nft) Init() error {
...
+		ipSet[net.IP(element.Key).String()] = struct{}{}
+	}
+	return ipSet
+}
+
+// remove duplicates, normalize decision timeouts
+func normalizedDecisions(decisions []*models.Decision) []*models.Decision {
+	vals := make(map[string]struct{})
+	finalDecisions := make([]*models.Decision, 0)
+	for _, d := range decisions {
```
Comment (buixor) [inline nftables.go:445]: as we saw, I think we should keep the longest valid decision for now (until LAPI can help us with that)

**Review by buixor — APPROVED** [review]
lgtm

**Review by buixor — APPROVED** [review]
great

**Pull request merged**